### PR TITLE
Make definition of readTQueue consistent with readTBQueue (#9539)

### PR DESCRIPTION
**Pull request by arybczak**
Details: https://ghc.haskell.org/trac/ghc/ticket/9539

Copy of my last comment:

Let's revive this ticket and get it resolved.

For what it's worth, readTBQueue uses lazy let for reverse and it's even appropriately commented: `NB. lazy: we want the transaction to be short, otherwise it will conflict.`

This issue makes me anxious to use TQueue in any real-world scenario because given sufficiently large spike in produced values it leads to memory exhaustion, as at some point the consumer will not be able to finish the transaction without conflicts even if rate of produced values goes down.

I'm just going to attach a patch that makes the definition of readTQueue consistent with readTBQueue and we can go from there.

If you know what exact benchmarks we need to run, please speak up.

-----

Basically the problem is as follows:
1) We know that current implementation of readTQueue causes issues.
2) It's inconsistent with appropriate function for TBQueue (that uses lazy binding).
3) Nobody knows why it's done the way it is (including Simon Marlow).

It's possible that current version is more efficient (but we need to measure it), but in the worst case I'd take an implementation that is somewhat less efficient over the one that might lead to memory exhaustion any day.

### CODE REVIEW

**Comment by bgamari** [comment]
> It's possible that current version is more efficient (but we need to measure it), but in the worst case I'd take an implementation that is somewhat less efficient over the one that might lead to memory exhaustion any day.

Yes, I agree. Slightly slower but predictable is certainly preferable to "fast but may blow up if you look at it wrong."
